import's are incorrectly generated if $GOPATH is on a symlink

## Problem statement

When the $GOPATH contains a symlink, the import paths are generated to locations using the realpath, rather than to the gopath location.

## Steps to reproduce


For example

```
export GOPATH=$HOME/Documents/Code/golang
```

```
 readlink "$HOME/Documents/Code"
/Users/billie/Dropbox/Documents/Code/
```

Then do your code generation

`swagger generate server swagger.json`

And your imports are generated relative to the realpath, rather than the gopath.

```go
import (
	"net/http"

	"github.com/go-openapi/runtime"

	"../../../../Dropbox/Documents/Code/golang/src/github.com/purplebooth/realsov/api/models"
)
```

When you compile 

```
go install ./cmd/real-sov-server
restapi/operations/authentication/get_oauth_token_responses.go:11:2: local import "../../../../Dropbox/Documents/Code/golang/src/github.com/purplebooth/realsov/api/models" in non-local package
```

## Environment

```
swagger version
version: 0.10.0
commit: 4d1e79b5a98c6f67bbf06923643d5317ed37a058
```

```
go version
go version go1.8.1 darwin/amd64
```

MacOS Sierra


## Workaround

Remove the symlink from your GOPATH. The imports are correctly generated.

--- a/.gitignore
+++ b/.gitignore
@@ -30,6 +30,8 @@ playground
 # Sublime files
 *.sublime-*
 
+# Vscode files
+.vscode/*
 cmd/swagger/swagger.json.bak
 tags
 

--- a/cmd/swagger/commands/generate/client.go
+++ b/cmd/swagger/commands/generate/client.go
@@ -102,14 +102,17 @@ func (c *Client) Execute(args []string) error {
 		return err
 	}
 
-	var basepath,rp string
+	var basepath,rp,targetAbs string
 
 	basepath,err = filepath.Abs(".")
 	if err != nil {
 		return err
 	}
-
-	rp, err = filepath.Rel(basepath, opts.Target)
+	targetAbs,err = filepath.Abs(opts.Target)
+	if err != nil {
+		return err
+	}
+		rp, err = filepath.Rel(basepath, targetAbs)
 	if err != nil {
 		return err
 	}

--- a/cmd/swagger/commands/generate/client.go
+++ b/cmd/swagger/commands/generate/client.go
@@ -102,7 +102,14 @@ func (c *Client) Execute(args []string) error {
 		return err
 	}
 
-	rp, err := filepath.Rel(".", opts.Target)
+	var basepath,rp string
+
+	basepath,err = filepath.Abs(".")
+	if err != nil {
+		return err
+	}
+
+	rp, err = filepath.Rel(basepath, opts.Target)
 	if err != nil {
 		return err
 	}

--- a/cmd/swagger/commands/generate/model.go
+++ b/cmd/swagger/commands/generate/model.go
@@ -91,7 +91,17 @@ func (m *Model) Execute(args []string) error {
 		return err
 	}
 
-	rp, err := filepath.Rel(".", opts.Target)
+	var basepath,rp,targetAbs string
+
+	basepath,err = filepath.Abs(".")
+	if err != nil {
+		return err
+	}
+	targetAbs,err = filepath.Abs(opts.Target)
+	if err != nil {
+		return err
+	}
+	rp, err = filepath.Rel(basepath, targetAbs)
 	if err != nil {
 		return err
 	}

--- a/cmd/swagger/commands/generate/operation.go
+++ b/cmd/swagger/commands/generate/operation.go
@@ -81,7 +81,17 @@ func (o *Operation) Execute(args []string) error {
 		return err
 	}
 
-	rp, err := filepath.Rel(".", opts.Target)
+	var basepath,rp,targetAbs string
+
+	basepath,err = filepath.Abs(".")
+	if err != nil {
+		return err
+	}
+	targetAbs,err = filepath.Abs(opts.Target)
+	if err != nil {
+		return err
+	}
+	rp, err = filepath.Rel(basepath, targetAbs)
 	if err != nil {
 		return err
 	}

--- a/cmd/swagger/commands/generate/server.go
+++ b/cmd/swagger/commands/generate/server.go
@@ -118,14 +118,17 @@ func (s *Server) Execute(args []string) error {
 	if e := generator.GenerateServer(s.Name, s.Models, s.Operations, opts); e != nil {
 		return e
 	}
-	var basepath, rp string
+	var basepath,rp,targetAbs string
 
 	basepath,err = filepath.Abs(".")
 	if err != nil {
 		return err
 	}
-
-	rp, err = filepath.Rel(basepath, opts.Target)
+	targetAbs,err = filepath.Abs(opts.Target)
+	if err != nil {
+		return err
+	}
+	rp, err = filepath.Rel(basepath, targetAbs)
 	if err != nil {
 		return err
 	}

--- a/cmd/swagger/commands/generate/server.go
+++ b/cmd/swagger/commands/generate/server.go
@@ -118,8 +118,14 @@ func (s *Server) Execute(args []string) error {
 	if e := generator.GenerateServer(s.Name, s.Models, s.Operations, opts); e != nil {
 		return e
 	}
+	var basepath, rp string
 
-	rp, err := filepath.Rel(".", opts.Target)
+	basepath,err = filepath.Abs(".")
+	if err != nil {
+		return err
+	}
+
+	rp, err = filepath.Rel(basepath, opts.Target)
 	if err != nil {
 		return err
 	}

--- a/cmd/swagger/commands/generate/support.go
+++ b/cmd/swagger/commands/generate/support.go
@@ -52,7 +52,17 @@ func (s *Support) Execute(args []string) error {
 		return err
 	}
 
-	rp, err := filepath.Rel(".", opts.Target)
+	var basepath,rp,targetAbs string
+	var err error
+	basepath,err = filepath.Abs(".")
+	if err != nil {
+		return err
+	}
+	targetAbs,err = filepath.Abs(opts.Target)
+	if err != nil {
+		return err
+	}
+	rp, err = filepath.Rel(basepath, targetAbs)
 	if err != nil {
 		return err
 	}

--- a/generator/support.go
+++ b/generator/support.go
@@ -173,7 +173,7 @@ func checkPrefixAndFetchRelativePath(childpath string, parentpath string) (bool,
 	}
 
 	return false, ""
-	
+
 
 }
 

--- a/generator/support.go
+++ b/generator/support.go
@@ -159,12 +159,30 @@ type appGenerator struct {
 	GenOpts         *GenOpts
 }
 
+// 1. Checks if the child path and parent path coincide.
+// 2. If they do return child path  relative to parent path.
+// 3. Everything else return false
+func checkPrefixAndFetchRelativePath(childpath string, parentpath string) (bool, string)  {
+
+	if strings.HasPrefix(childpath, parentpath) {
+		pth, err := filepath.Rel(parentpath, childpath)
+		if err != nil {
+			log.Fatalln(err)
+		}
+		return true, pth
+	} else {
+		return false, ""
+	}
+
+}
+
 func baseImport(tgt string) string {
 	tgtAbsPath, err := filepath.Abs(tgt)
 	if err != nil {
 		log.Fatalln(err)
 	}
-	tgtAbsPath,err = filepath.EvalSymlinks(tgtAbsPath)
+	var tgtAbsPathExtended string
+	tgtAbsPathExtended,err = filepath.EvalSymlinks(tgtAbsPath)
 	if err != nil {
 		log.Fatalln(err)
 	}
@@ -182,31 +200,60 @@ func baseImport(tgt string) string {
 			log.Fatalln(err)
 		}
 		gopathExtended = filepath.Join(gopathExtended, "src")
-
-		// At this stage both target path and GOPATH are
-		// absolutely expanded and symlink free paths.
-		// This will ensure correct path generation below.
+		gp = filepath.Join(gp,"src")
 
 		// Windows (local) file systems - NTFS, as well as FAT and variants
 		// are case insensitive.
 		if goruntime.GOOS == "windows" {
 			tgtAbsPath = strings.ToLower(tgtAbsPath)
+			tgtAbsPathExtended = strings.ToLower(tgtAbsPathExtended)
 			gopathExtended = strings.ToLower(gopathExtended)
+			gp = strings.ToLower(gp)
 		}
 
-		if strings.HasPrefix(tgtAbsPath, gopathExtended) {
-			pth, err = filepath.Rel(gopathExtended, tgtAbsPath)
-			if err != nil {
-				log.Fatalln(err)
-			}
+		// At this stage we have expanded and unexpanded target path. GOPATH is fully expanded.
+		// Expanded means symlink free.
+		// We compare both types of targetpath<s> with gopath.
+		// If any one of them coincides with gopath , it is imperative that
+		// target path lies inside gopath. How?
+		// 		- Case 1: Irrespective of symlinks paths coincide. Both non-expanded paths.
+		// 		- Case 2: Symlink in target path points to location inside GOPATH. (Expanded Target Path)
+		//    - Case 3: Symlink in target path points to directory outside GOPATH (Unexpanded target path)
+
+		// Case 1: - Do nothing case. If non-expanded paths match just genrate base import path as if
+		//				   there are no symlinks.
+
+		// Case 2: - Symlink in target path points to location inside GOPATH. (Expanded Target Path)
+		//					 First if will fail. Second if will succeed.
+
+		// Case 3: - Symlink in target path points to directory outside GOPATH (Unexpanded target path)
+		// 					 First if will succeed and break.
+
+
+		//compares non expanded path for both
+		if ok,relativepath := checkPrefixAndFetchRelativePath(tgtAbsPath, gp); ok {
+			pth = relativepath
 			break
 		}
+
+		// Compares non-expanded target path
+		if ok,relativepath := checkPrefixAndFetchRelativePath(tgtAbsPath, gopathExtended); ok {
+			pth = relativepath
+			break
+		}
+
+		// Compares expanded target path.
+		if ok,relativepath := checkPrefixAndFetchRelativePath(tgtAbsPathExtended, gopathExtended); ok {
+			pth = relativepath
+			break
+		}
+
 	}
 
 	if pth == "" {
 		log.Fatalln("target must reside inside a location in the $GOPATH/src")
 	}
-	return pth
+ 	return pth
 }
 
 func (a *appGenerator) Generate() error {

--- a/generator/support.go
+++ b/generator/support.go
@@ -170,9 +170,10 @@ func checkPrefixAndFetchRelativePath(childpath string, parentpath string) (bool,
 			log.Fatalln(err)
 		}
 		return true, pth
-	} else {
-		return false, ""
 	}
+
+	return false, ""
+	
 
 }
 

--- a/generator/support.go
+++ b/generator/support.go
@@ -160,7 +160,11 @@ type appGenerator struct {
 }
 
 func baseImport(tgt string) string {
-	p, err := filepath.Abs(tgt)
+	tgtAbsPath, err := filepath.Abs(tgt)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	tgtAbsPath,err = filepath.EvalSymlinks(tgtAbsPath)
 	if err != nil {
 		log.Fatalln(err)
 	}
@@ -172,14 +176,26 @@ func baseImport(tgt string) string {
 
 	var pth string
 	for _, gp := range filepath.SplitList(gopath) {
-		pp := filepath.Join(filepath.Clean(gp), "src")
-		var np, npp string
-		if goruntime.GOOS == "windows" {
-			np = strings.ToLower(p)
-			npp = strings.ToLower(pp)
+		// EvalSymLinks also calls the Clean
+		gopathExtended, err := filepath.EvalSymlinks(gp)
+		if err != nil {
+			log.Fatalln(err)
 		}
-		if strings.HasPrefix(np, npp) {
-			pth, err = filepath.Rel(pp, p)
+		gopathExtended = filepath.Join(gopathExtended, "src")
+
+		// At this stage both target path and GOPATH are
+		// absolutely expanded and symlink free paths.
+		// This will ensure correct path generation below.
+
+		// Windows (local) file systems - NTFS, as well as FAT and variants
+		// are case insensitive.
+		if goruntime.GOOS == "windows" {
+			tgtAbsPath = strings.ToLower(tgtAbsPath)
+			gopathExtended = strings.ToLower(gopathExtended)
+		}
+
+		if strings.HasPrefix(tgtAbsPath, gopathExtended) {
+			pth, err = filepath.Rel(gopathExtended, tgtAbsPath)
 			if err != nil {
 				log.Fatalln(err)
 			}

--- a/generator/support_test.go
+++ b/generator/support_test.go
@@ -5,6 +5,7 @@ import (
 	"os"
 	goruntime "runtime"
 	"path/filepath"
+	"strings"
 )
 
 var checkprefixandfetchrelativepathtests = []struct {
@@ -47,13 +48,14 @@ var checkbaseimporttest = []struct {
 
 }
 
+
 func TestCheckPrefixFetchRelPath(t *testing.T) {
 
 	for _,item := range checkprefixandfetchrelativepathtests {
 		actualok, actualpath := checkPrefixAndFetchRelativePath(item.childpath, item.parentpath)
 
 		if goruntime.GOOS == "windows" {
-			item.path = filepath.Clean(item.path)
+			item.path = strings.Replace(item.path, "/", "\\", -1)
 		}
 
 		if actualok != item.ok {
@@ -97,8 +99,7 @@ func TestBaseImport(t *testing.T) {
 		actualpath := baseImport(item.targetpath)
 
 		if goruntime.GOOS == "windows" {
-			actualpath = filepath.Clean(actualpath)
-			item.expectedpath = filepath.Clean(item.expectedpath)
+			item.expectedpath = strings.Replace(item.expectedpath, "/", "\\", -1)
 		}
 
 		if actualpath != item.expectedpath {

--- a/generator/support_test.go
+++ b/generator/support_test.go
@@ -3,6 +3,8 @@ package generator
 import (
 	"testing"
 	"os"
+	goruntime "runtime"
+	"path/filepath"
 )
 
 var checkprefixandfetchrelativepathtests = []struct {
@@ -49,6 +51,11 @@ func TestCheckPrefixFetchRelPath(t *testing.T) {
 
 	for _,item := range checkprefixandfetchrelativepathtests {
 		actualok, actualpath := checkPrefixAndFetchRelativePath(item.childpath, item.parentpath)
+
+		if goruntime.GOOS == "windows" {
+			actualpath = filepath.Clean(actualpath)
+		}
+
 		if actualok != item.ok {
 			t.Errorf("checkPrefixAndFetchRelativePath(%s, %s): expected %v, actual %v", item.childpath, item.parentpath, item.ok, actualok)
 		} else if actualpath != item.path {
@@ -88,6 +95,11 @@ func TestBaseImport(t *testing.T) {
 
 		// Test
 		actualpath := baseImport(item.targetpath)
+
+		if goruntime.GOOS == "windows" {
+			actualpath = filepath.Clean(actualpath)
+			item.expectedpath = filepath.Clean(item.expectedpath)
+		}
 
 		if actualpath != item.expectedpath {
 			t.Errorf("baseImport(%s): expected %s, actual %s", item.targetpath, item.expectedpath, actualpath)

--- a/generator/support_test.go
+++ b/generator/support_test.go
@@ -1,17 +1,17 @@
 package generator
 
 import (
-	"testing"
 	"os"
 	goruntime "runtime"
 	"strings"
+	"testing"
 )
 
 var checkprefixandfetchrelativepathtests = []struct {
 	childpath  string
 	parentpath string
-	ok bool
-	path string
+	ok         bool
+	path       string
 }{
 	// Positive
 	{"/", "/", true, "."},
@@ -23,34 +23,31 @@ var checkprefixandfetchrelativepathtests = []struct {
 	{"/User/Gopher", "/User/SomethingElse", false, ""},
 	{"/var", "/etc", false, ""},
 	{"/mnt/dev3", "/mnt/dev3/dir", false, ""},
-
-
-
 }
 
+var tempdir = os.TempDir()
+
 var checkbaseimporttest = []struct {
-	path  []string
-	gopath string
-	targetpath string
-	symlinksrc string
-	symlinkdest string  // symlink is the last dir in targetpath
+	path         []string
+	gopath       string
+	targetpath   string
+	symlinksrc   string
+	symlinkdest  string // symlink is the last dir in targetpath
 	expectedpath string
 }{
 	// No sym link. Positive Test Case
-	{[]string {"/tmp/root/go/src/github.com/go-swagger",}, "/tmp/root/go/", "/tmp/root/go/src/github.com/go-swagger", "", "", "github.com/go-swagger"},
+	{[]string{tempdir + "/root/go/src/github.com/go-swagger"}, tempdir + "/root/go/", tempdir + "/root/go/src/github.com/go-swagger", "", "", "github.com/go-swagger"},
 	// Symlink points inside GOPATH
-	{[]string {"/tmp/root/go/src/github.com/go-swagger",}, "/tmp/root/go/", "/tmp/root/symlink", "/tmp/root/symlink", "/tmp/root/go/src/", "."},
+	{[]string{tempdir + "/root/go/src/github.com/go-swagger"}, tempdir + "/root/go/", tempdir + "/root/symlink", tempdir + "/root/symlink", tempdir + "/root/go/src/", "."},
 	// Symlink points inside GOPATH
-	{[]string {"/tmp/root/go/src/github.com/go-swagger",}, "/tmp/root/go/", "/tmp/root/symlink", "/tmp/root/symlink", "/tmp/root/go/src/github.com", "github.com"},
+	{[]string{tempdir + "/root/go/src/github.com/go-swagger"}, tempdir + "/root/go/", tempdir + "/root/symlink", tempdir + "/root/symlink", tempdir + "/root/go/src/github.com", "github.com"},
 	// Symlink point outside GOPATH : Targets Case 1: in baseImport implementation.
-	{[]string {"/tmp/root/go/src/github.com/go-swagger","/tmp/root/gopher/go/"}, "/tmp/root/go/", "/tmp/root/go/src/github.com/gopher", "/tmp/root/go/src/github.com/gopher", "/tmp/root/gopher/go", "github.com/gopher"},
-
+	{[]string{tempdir + "/root/go/src/github.com/go-swagger", tempdir + "/root/gopher/go/"}, tempdir + "/root/go/", tempdir + "/root/go/src/github.com/gopher", tempdir + "/root/go/src/github.com/gopher", tempdir + "/root/gopher/go", "github.com/gopher"},
 }
-
 
 func TestCheckPrefixFetchRelPath(t *testing.T) {
 
-	for _,item := range checkprefixandfetchrelativepathtests {
+	for _, item := range checkprefixandfetchrelativepathtests {
 		actualok, actualpath := checkPrefixAndFetchRelativePath(item.childpath, item.parentpath)
 
 		if goruntime.GOOS == "windows" {
@@ -79,12 +76,12 @@ func TestBaseImport(t *testing.T) {
 
 	oldgopath := os.Getenv("GOPATH")
 	defer os.Setenv("GOPATH", oldgopath)
-	defer os.RemoveAll("/tmp/root")
+	defer os.RemoveAll(tempdir + "/root")
 
-	for _,item := range checkbaseimporttest {
+	for _, item := range checkbaseimporttest {
 
 		// Create Paths
-		for _,paths := range item.path {
+		for _, paths := range item.path {
 			os.MkdirAll(paths, 0777)
 		}
 
@@ -105,7 +102,7 @@ func TestBaseImport(t *testing.T) {
 			t.Errorf("baseImport(%s): expected %s, actual %s", item.targetpath, item.expectedpath, actualpath)
 		}
 
-		os.RemoveAll("/tmp/root")
+		os.RemoveAll(tempdir + "/root")
 
 	}
 

--- a/generator/support_test.go
+++ b/generator/support_test.go
@@ -4,7 +4,6 @@ import (
 	"testing"
 	"os"
 	goruntime "runtime"
-	"path/filepath"
 	"strings"
 )
 

--- a/generator/support_test.go
+++ b/generator/support_test.go
@@ -1,0 +1,100 @@
+package generator
+
+import (
+	"testing"
+	"os"
+)
+
+var checkprefixandfetchrelativepathtests = []struct {
+	childpath  string
+	parentpath string
+	ok bool
+	path string
+}{
+	// Positive
+	{"/", "/", true, "."},
+	{"/User/Gopher", "/", true, "User/Gopher"},
+	{"/User/Gopher/Go", "/User/Gopher/Go", true, "."},
+	{"/User/../User/Gopher", "/", true, "User/Gopher"},
+	// Negative cases
+	{"/", "/var", false, ""},
+	{"/User/Gopher", "/User/SomethingElse", false, ""},
+	{"/var", "/etc", false, ""},
+	{"/mnt/dev3", "/mnt/dev3/dir", false, ""},
+
+
+
+}
+
+var checkbaseimporttest = []struct {
+	path  []string
+	gopath string
+	targetpath string
+	symlinksrc string
+	symlinkdest string  // symlink is the last dir in targetpath
+	expectedpath string
+}{
+	// No sym link. Positive Test Case
+	{[]string {"/tmp/root/go/src/github.com/go-swagger",}, "/tmp/root/go/", "/tmp/root/go/src/github.com/go-swagger", "", "", "github.com/go-swagger"},
+	// Symlink points inside GOPATH
+	{[]string {"/tmp/root/go/src/github.com/go-swagger",}, "/tmp/root/go/", "/tmp/root/symlink", "/tmp/root/symlink", "/tmp/root/go/src/", "."},
+	// Symlink points inside GOPATH
+	{[]string {"/tmp/root/go/src/github.com/go-swagger",}, "/tmp/root/go/", "/tmp/root/symlink", "/tmp/root/symlink", "/tmp/root/go/src/github.com", "github.com"},
+	// Symlink point outside GOPATH : Targets Case 1: in baseImport implementation.
+	{[]string {"/tmp/root/go/src/github.com/go-swagger","/tmp/root/gopher/go/"}, "/tmp/root/go/", "/tmp/root/go/src/github.com/gopher", "/tmp/root/go/src/github.com/gopher", "/tmp/root/gopher/go", "github.com/gopher"},
+
+}
+
+func TestCheckPrefixFetchRelPath(t *testing.T) {
+
+	for _,item := range checkprefixandfetchrelativepathtests {
+		actualok, actualpath := checkPrefixAndFetchRelativePath(item.childpath, item.parentpath)
+		if actualok != item.ok {
+			t.Errorf("checkPrefixAndFetchRelativePath(%s, %s): expected %v, actual %v", item.childpath, item.parentpath, item.ok, actualok)
+		} else if actualpath != item.path {
+			t.Errorf("checkPrefixAndFetchRelativePath(%s, %s): expected %s, actual %s", item.childpath, item.parentpath, item.path, actualpath)
+		} else {
+			continue
+		}
+	}
+
+}
+
+func TestBaseImport(t *testing.T) {
+
+	// 1. Create a root folder /tmp/root
+	// 2. Simulate scenario
+	//	2.a No Symlink
+	//	2.b Symlink from outside of GOPATH to inside
+	//  2.c Symlink from inside of GOPATH to outside.
+	// 3. Check results.
+
+	oldgopath := os.Getenv("GOPATH")
+	defer os.Setenv("GOPATH", oldgopath)
+	defer os.RemoveAll("/tmp/root")
+
+	for _,item := range checkbaseimporttest {
+
+		// Create Paths
+		for _,paths := range item.path {
+			os.MkdirAll(paths, 0777)
+		}
+
+		// Change GOPATH
+		os.Setenv("GOPATH", item.gopath)
+
+		// Create Symlink
+		os.Symlink(item.symlinkdest, item.symlinksrc)
+
+		// Test
+		actualpath := baseImport(item.targetpath)
+
+		if actualpath != item.expectedpath {
+			t.Errorf("baseImport(%s): expected %s, actual %s", item.targetpath, item.expectedpath, actualpath)
+		}
+
+		os.RemoveAll("/tmp/root")
+
+	}
+
+}

--- a/generator/support_test.go
+++ b/generator/support_test.go
@@ -53,7 +53,7 @@ func TestCheckPrefixFetchRelPath(t *testing.T) {
 		actualok, actualpath := checkPrefixAndFetchRelativePath(item.childpath, item.parentpath)
 
 		if goruntime.GOOS == "windows" {
-			actualpath = filepath.Clean(actualpath)
+			item.path = filepath.Clean(item.path)
 		}
 
 		if actualok != item.ok {

